### ISSUE TITLE
Pod startup fails when  /var/lib/docker/containers is a separate file system

### ISSUE THREAD

**Issue by ansilh** [rancher/rancher#29275]
**What kind of request is this (question/bug/enhancement/feature request):**
bug/enhancement 

**Steps to reproduce (least amount of steps as possible):**

Created a separate file system for /var/lib/docker/containers and deploy rke k8s cluster

**Result:**

Pod startups are failing and kubelet shows below error in the logs.
```
"/var/lib/docker/containers/86eae9a7c4508413f8e40f811e29bbad568c3008b0a14195b46223051df6337e/resolv.conf" does not exist
38s Warning FailedCreatePodSandBox pod/busyboxplus-vm9c9 (combined from similar events): Failed create pod sandbox: rpc error: code = Unknown desc = rewrite resolv.conf failed for pod "busyboxplus-vm9c9": ResolvConfPath "/var/lib/docker/containers/9294866d84f2a315d9d4e8bc2b4c524cbad52082136a12c2226c70f157ddcd55/resolv.conf" does not exist
```
**Other details that may be helpful:**

FS mount flags: `https://github.com/rancher/rke/blob/master/cluster/plan.go#L477`

```
host.DockerInfo.DockerRootDir + ":" + host.DockerInfo.DockerRootDir + ":rw,rslave,z",
```

Flag meanings;
```
rw     - Mount as read-write 
rslave - Sub-mounts of the original mount are exposed to replica mounts \
         the propagation also extends to and from mount points nested within any of the \
         original or replica mount points
z      - Option tells Docker that the volume content will be shared between containers
```

Ideally , the mount `/var/lib/docker` and the submount `/var/lib/docker/containers` should be visible inside kubelet container. (Tested these flags on a buysbox container and noticed that submounts insides /var/lib/docker are visible)
Result:
 File system `/var/lib/docker/containers` is not visible inside kubelet container
While checking the kubelet logs ,found that entrypoint.sh is doing unmount operation during the startup

```
$ docker logs kubelet 2>&1 |grep umount
...
+ '[' /var/lib/docker/containers '!=' /run/nscd ']'
+ umount /var/lib/docker/containers
```

We need some logic in the script to exclude  `/var/lib/docker/containers`  while unmounting the subdirectories if this one is a separate mount point

**Cluster information**
- Cluster type : NA
- Machine type : NA
- Kubernetes version:

```
Client Version: version.Info{Major:"1", Minor:"18", GitVersion:"v1.18.6", GitCommit:"dff82dc0de47299ab66c83c626e08b245ab19037", GitTreeState:"clean", BuildDate:"2020-07-15T16:58:53Z", GoVersion:"go1.13.9", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"18", GitVersion:"v1.18.6", GitCommit:"dff82dc0de47299ab66c83c626e08b245ab19037", GitTreeState:"clean", BuildDate:"2020-07-15T16:51:04Z", GoVersion:"go1.13.9", Compiler:"gc", Platform:"linux/amd64"}
```

- Docker version (use `docker version`):

```
Client: Docker Engine - Community
 Version:           19.03.12
 API version:       1.40
 Go version:        go1.13.10
 Git commit:        48a66213fe
 Built:             Mon Jun 22 15:45:36 2020
 OS/Arch:           linux/amd64
 Experimental:      false

Server: Docker Engine - Community
 Engine:
  Version:          19.03.12
  API version:      1.40 (minimum version 1.12)
  Go version:       go1.13.10
  Git commit:       48a66213fe
  Built:            Mon Jun 22 15:44:07 2020
  OS/Arch:          linux/amd64
  Experimental:     false
 containerd:
  Version:          1.3.7
  GitCommit:        8fba4e9a7d01810a393d5d25a3621dc101981175
 runc:
  Version:          1.0.0-rc10
  GitCommit:        dc9208a3303feef5b3839f4323d9beb36df0a9dd
 docker-init:
  Version:          0.18.0
  GitCommit:        fec3683
```

gz#12051

**Comment by sadiapoddar** [comment]
Reproduced the issue with the help and steps provided by @ansilh 
Steps to repro.
Here are the steps I’ve followed

1. Created 2 ec2 instances
2. Created a rke cluster Rancher UI and added those two nodes in the cluster( one worker, other etcd+cntrl)
3. Made sure the cluster is up and running. 
4. Stopped docker on both instances 

Creating a new file system to bring the content from /var/lib/docker/containers 

1. Stopped  docker on both VMs
2. Added new disk to the worker VM
3. Created a new vg/logical volume using the new volume and mounted it over /mnt.

>vgcreate vgDocker /dev/nvme1n1
>lvcreate -n lvCtrs -L+10G vgDocker
>mkfs.ext4 /dev/vgDocker/lvCtrs
>mount /dev/vgDocker/lvCtrs /mnt
>cp -prv /var/lib/docker/containers/* /mnt/.
4. Unmounted /mnt and mounted the file system on top of /var/lib/docker/containers/
5. Started docker on both vms

**Errors in log show pod start is failing**
------
E1113 05:42:09.918016    5176 remote_runtime.go:113] RunPodSandbox from runtime service failed: rpc error: code = Unknown desc = rewrite resolv.conf failed for pod "coredns-6f85d5fb88-m7l8j": ResolvConfPath "/var/lib/docker/containers/fe5e7c25528e2af8444f8c089d3644b02fd66b342009f77d8f347e9f41dee5fd/resolv.conf" does not exist
E1113 05:42:09.918162    5176 kuberuntime_sandbox.go:69] CreatePodSandbox for pod "coredns-6f85d5fb88-m7l8j_kube-system(e269041a-516e-4f8d-acaf-54d5be8fbbda)" failed: rpc error: code = Unknown desc = rewrite resolv.conf failed for pod "coredns-6f85d5fb88-m7l8j": ResolvConfPath "/var/lib/docker/containers/fe5e7c25528e2af8444f8c089d3644b02fd66b342009f77d8f347e9f41dee5fd/resolv.conf" does not exist
E1113 05:42:09.918179    5176 kuberuntime_manager.go:730] createPodSandbox for pod "coredns-6f85d5fb88-m7l8j_kube-system(e269041a-516e-4f8d-acaf-54d5be8fbbda)" failed: rpc error: code = Unknown desc = rewrite resolv.conf failed for pod "coredns-6f85d5fb88-m7l8j": ResolvConfPath "/var/lib/docker/containers/fe5e7c25528e2af8444f8c089d3644b02fd66b342009f77d8f347e9f41dee5fd/resolv.conf" does not exist

**Comment by sadiapoddar** [comment]
Verified 2.5-head, v2.5-ef237669,

1. Created 2 ec2 instances
2. Created a rke cluster Rancher UI and added those two nodes in the cluster( one worker, other etcd+cntrl)
3. Made sure the cluster is up and running.
4. Stopped docker on both instances
5. Creating a new file system to bring the content from /var/lib/docker/containers
6. Stopped docker on both VMs
7. Added new disk to the worker VM
8. Created a new vg/logical volume using the new volume and mounted it over /mnt.
9. vgcreate vgDocker /dev/nvme1n1
10. lvcreate -n lvCtrs -L+10G vgDocker
11. mkfs.ext4 /dev/vgDocker/lvCtrs
12. mount /dev/vgDocker/lvCtrs /mnt
13. cp -prv /var/lib/docker/containers/* /mnt/.
14. Unmounted /mnt and mounted the file system on top of /var/lib/docker/containers/
15. Started docker on both vms,

The filesystem was mounted properly 
<img width="781" alt="Screen Shot 2020-12-05 at 2 47 20 PM" src="https://user-images.githubusercontent.com/72409237/101267032-7a3d4080-3709-11eb-8c1e-94a61fdc912a.png">


Not seeing the same startup error but saw some other errors at the starting time, 
<img width="1820" alt="Screen Shot 2020-12-05 at 2 54 32 PM" src="https://user-images.githubusercontent.com/72409237/101267233-5d097180-370b-11eb-9038-182f1c8ccb77.png">

The cluster is up and running, kubelet in workers are green. 
<img width="1694" alt="Screen Shot 2020-12-05 at 3 02 12 PM" src="https://user-images.githubusercontent.com/72409237/101267175-e2d8ed00-370a-11eb-9c07-dfd1b2cb67b8.png">
 
I am keeping the setup and will sync-up with @paynejacob and @ansilh to verify if those errors are related to the fix.

**Comment by sadiapoddar** [comment]
Reopening the issue for @paynejacob to confirm if the errors during staring kublets are not any concern. 

<img width="1820" alt="101267233-5d097180-370b-11eb-9038-182f1c8ccb77" src="https://user-images.githubusercontent.com/72409237/101390596-f05fb580-3877-11eb-971f-deddc455dc09.png">

**Comment by paynejacob** [comment]
@sadiapoddar When testing this can we change the docker mount before installing rke?

**Comment by sadiapoddar** [comment]
@paynejacob You are suggesting we should mount the /var/lib/docker/containers over the added volume before installing the rancher?

**Comment by paynejacob** [comment]
Yes, when I do it in that order it works fine.

**Comment by sadiapoddar** [comment]
@paynejacob, Then this issue is not fixed? Since this issue was the filesystem was mounted after the rancher installation won't be supported now?

**Comment by paynejacob** [comment]
That is a good point, ill see if I can find a solution for that case.

**Comment by sadiapoddar** [comment]
@paynejacob Tested the following workflow. disk added before creating a rancher cluster. 
1. created two ec2, and added volume on one of the ec2s.
2. stopped docker and mounted /var/lib/docker/containters on the added volume as a mount point
3. created a custom cluster by adding the mounted volume ec2 as the worker node
4. Cluster came up fine.
5. Didn't find the "rootDisk" and "extraDisk" error in worker node kubelet logs

Reopening for @paynejacob to investigate if we can support disk mount on worker node on the existing cluster.

**Comment by paynejacob** [comment]
@sadiapoddar I tried to reproduce this with the following steps and was unable to.  Can you please try to reproduce again?

1. install docker on a do droplet
2. rke up with a single node
3. stop docker service docker stop
4. mount /media/VHD.img /mnt/VHD
5. cp -prvf /var/lib/docker/container /mnt/VHD
6. rm -rf/var/lib/docker/container/*
7. umount /mnt/VHD
8. mount /media/VHD.img /var/lib/docker/container
9. service docker start

**Comment by sadiapoddar** [comment]
@paynejacob 
Tested on the latest master branch, couldn't find the "rootDisk" and "extraDisk" in kubelet log.
I did the exact same steps as before but looks like the issue is fixed. 

Steps used to verify. 

- Created 2 ec2 instances
- Created a rke cluster Rancher UI and added those two nodes in the cluster( one worker, other etcd+cntrl)
- Made sure the cluster is up and running.
- Stopped docker on both instances
- Creating a new file system to bring the content from /var/lib/docker/containers
- Stopped docker on both VMs
- Added new disk to the worker VM
- Created a new vg/logical volume using the new volume and mounted it over /mnt.
- vgcreate vgDocker /dev/nvme1n1
- lvcreate -n lvCtrs -L+10G vgDocker
- mkfs.ext4 /dev/vgDocker/lvCtrs
- mount /dev/vgDocker/lvCtrs /mnt
- cp -prv /var/lib/docker/containers/* /mnt/.
- Unmounted /mnt and mounted the file system on top of /var/lib/docker/containers/
- Started docker on both vms,

**Comment by sadiapoddar** [comment]
Tested the fix with 2.5.4-rc5.
Looks like the fix is working for both the use cases.
1. Adding extra disk and mounting /var/lib/docker/containters  on that extra disk  before creating the rancher cluster
2. Adding extra disk and mounting /var/lib/docker/containters  on that extra disk after creating the rancher cluster
@paynejacob  Thanks for fixing this issue.
